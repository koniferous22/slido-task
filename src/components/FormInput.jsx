import React from 'react'

const FormInput = props => {
	const {name, label, errorFlag, onChange, onBlur, ...rest} = props	
	return (
		<>
			{
				errorFlag && (
					<div className="error">
						Field is required
					</div>
				)
			}
			<label htmlFor={name}>
				{label + ': '}
			</label>
			<input
				{...rest}
				name={name}
				onChange={(event) => onChange(name, event)}
				onBlur={() => onBlur(name)}
			/>			
		</>
	)
}

export default FormInput
