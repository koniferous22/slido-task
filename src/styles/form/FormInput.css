form#case-add input {
	border-radius: 5px;
}