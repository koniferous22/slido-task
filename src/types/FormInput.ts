export interface FormInputProps {
	name: string,
	label: string,
	type: string,
	value: string,
	errorFlag: boolean,
	onBlur: (s: string) => void,
	onChange: (s: string, e: React.ChangeEvent<HTMLInputElement>) => void
}
