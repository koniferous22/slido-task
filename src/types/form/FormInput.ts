import { Case } from 'types/common/Case'

export interface FormInputProps {
	name: keyof Case;
	label: string;
	type: string;
	value: string;
	errorFlag: boolean;
	onBlur: (field: keyof Case) => void;
	onChange: (field: keyof Case, value: string) => void;
}
