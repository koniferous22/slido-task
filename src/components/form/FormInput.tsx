import React from 'react'
import { FormInputProps } from 'types/form/FormInput'

import 'styles/form/FormInput.css'

const FormInput: React.FC<FormInputProps> = ({name, label, errorFlag, onChange, onBlur, ...inputProps}: FormInputProps) => {
	return (
		<>
			{
				errorFlag && (
					<div className="error" id={"error-" + name}>
						Field is required
					</div>
				)
			}
			<label htmlFor={name}>
				{label + ': '}
			</label>
			<input
				{...inputProps}
				id={name}
				name={name}
				onChange={(event): void => onChange(name, event.target.value)}
				onBlur={(): void => onBlur(name)}
			/>			
		</>
	)
}

export default FormInput
