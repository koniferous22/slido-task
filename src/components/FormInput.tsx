import React from 'react'
import { FormInputProps } from '../types/FormInput'

const FormInput = (props: FormInputProps) => {
	const {name, label, errorFlag, onChange, onBlur, ...rest} = props	
	return (
		<>
			{
				errorFlag && (
					<div className="error">
						Field is required
					</div>
				)
			}
			<label htmlFor={name}>
				{label + ': '}
			</label>
			<input
				{...rest}
				name={name}
				onChange={(event) => onChange(name, event)}
				onBlur={() => onBlur(name)}
			/>			
		</>
	)
}

export default FormInput
