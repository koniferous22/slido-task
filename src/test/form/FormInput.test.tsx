import React from 'react'
import renderer from 'react-test-renderer'
import FormInput from 'components/form/FormInput'

import { cleanup, fireEvent, render } from '@testing-library/react'

afterEach(cleanup)

it('FormInput renders correctly', () => {
	const tree = renderer
		.create(
			<FormInput
				name="title"
				type="text"
				label="Input"
				value="sample value"
				errorFlag={false}
				onChange={(): void => {return;}}
				onBlur={(): void => {return;}}
			/>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('FormInput renders correctly with error', () => {
	const tree = renderer
		.create(
			<FormInput
				name="date"
				type="text"
				label="Input"
				value="sample value"
				errorFlag={true}
				onChange={(): void => {return;}}
				onBlur={(): void => {return;}}
			/>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

it('onchange, onblur mocks', () => {
	const mockOnChange = jest.fn()
	const mockOnBlur = jest.fn()
	const { getByLabelText } = render(<FormInput
		name="date"
		type="text"
		label="Input"
		value="sample value"
		errorFlag={true}
		onChange={mockOnChange}
		onBlur={mockOnBlur}
	/>)
	fireEvent.change(getByLabelText(/Input/), { target: { value: 'a' } });
	expect(mockOnChange).toHaveBeenCalled()
	fireEvent.blur(getByLabelText(/Input/));
	expect(mockOnBlur).toHaveBeenCalled()
})