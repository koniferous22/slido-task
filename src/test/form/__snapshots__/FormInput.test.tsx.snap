// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput renders correctly 1`] = `
Array [
  <label
    htmlFor="title"
  >
    Input: 
  </label>,
  <input
    id="title"
    name="title"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="sample value"
  />,
]
`;

exports[`FormInput renders correctly with error 1`] = `
Array [
  <div
    className="error"
    id="error-date"
  >
    Field is required
  </div>,
  <label
    htmlFor="date"
  >
    Input: 
  </label>,
  <input
    id="date"
    name="date"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="sample value"
  />,
]
`;
